[FLINK-2405] [streaming] Added stateful functions to Scala DataStreams

gyfora opened the pull request:
This PR introduces 3 new API methods for the Scala DataStreams that extend the standard Map, FlatMap, and Filter function with state.

**These methods are:**
*for a DataStream[T]*
* **mapWithState**(fun: (T, Option[S]) => (R, Option[S])): DataStream[R]
* **flatMapWithState**( fun: (T, Option[S]) => (TraversableOnce[R], Option[S])): DataStream[R]
* **filterWithState**(fun: (T, Option[S]) => (Boolean, Option[S])): DataStream[T]

State partitioning is applied whenever the transformations are applied on a KeyedDataStream. Otherwise local state is used. Please refer to these concepts in:
https://cwiki.apache.org/confluence/display/FLINK/Stateful+Stream+Processing.

I also added an IT case to check the functionality

aljoscha commented:
Changes look good. The only thing that might seem (aesthetically) problematic is that you cannot use the nice no-parenthesis syntax for lambdas if you want to specify the stateByKey parameter. You have to use this:
doIt({ a => a + 3 }, false)

Maybe @tillrohrmann has a suggestion.

gyfora commented:
I agree that it's not optimal, but I couldn't figure out a better way, let's here Till :)
Btw I think you can write: doIt( a => a + 3 , false) so its not all that bad

tillrohrmann commented on flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/function/StatefulFunction.scala:
```diff
@@ -0,0 +1,42 @@
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.api.common.state.OperatorState
+
+trait StatefulFunction[I, O, S] extends RichFunction {
```
Scala docs are missing here.

tillrohrmann commented on flink-staging/flink-streaming/flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala:
```diff
@@ -332,13 +359,23 @@ class DataStreamTest {
+        (in, state: Option[Long]) => (false, None), true)
+    assert(getFunctionForDataStream(statefulFilter2).
+        asInstanceOf[StatefulFunction[_, _, _]].isPartitioned)
+   
```
Maybe we can add an IT test which is stateful and uses the new API constructs. This will assure that the values are correctly read from and written to the `OperatorState`.

gyfora replied:
Alright, I will do that. :+1:

tillrohrmann commented:
I don't know of a better solution than the current approach if we want to support the `stateByKey` parameter. I agree with Aljoscha that `ds.mapWithState({ (x, o) => (x, o) }, true)` does not look really nice and you will always have it if you have a multi line UDF.

I was wondering whether we can get rid of the `stateByKey` parameter. Is there any use case in which you have a `KeyedDataStream` and set `stateByKey` to false? Otherwise we could override the `*WithState` methods in `KeyedDataStream` and per default the `StatefulFunction` sets the `partitioned` field to `false`?

Other than that, the PR looks really good and I like it a lot :-)

gyfora commented:
You actually make a good point. Since one cannot apply multiple transformations while keeping a KeyedDataStream, we can assume stateByKey true on a keyedDataStream, otherwise it does not make sense to call keyBy before it.

I modify it.

gyfora commented:
I updated the PR according to the comments, and added an IT case that tests all 3 functions. (map-localstate, flatmap-localstate, filter-partitionedstate)

Let's see if travis likes it or not :)

gyfora commented:
Travis apparently passes (there was some unrelated Kafka error), so if no objections I will merge this tomorrow :)

aljoscha commented:
Sure, go ahead. :smile:

tillrohrmann commented:
I've got one remark before merging.

tillrohrmann commented:
I think the way the partitioned field of a `StatefulFunction` is set can be simplified. Currently this happens with via the partitionByStateKey. This is not very intuitive. It would be easier IMHO to set it directly in the class instantiation. I opened a PR against Gyula's PR branch.

gyfora commented:
Thanks, I will check :)
